File: array/max_erasure_value.cpp
//https://leetcode.com/problems/maximum-erasure-value/
class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int l = 0, r = 0, ans = 0, n = nums.size(), sum = 0;
        unordered_map<int, int> last_occ; //value to index
        for(; r < n; r++)
        {
            int r_val = r_val;
            sum += r_val;
            //if r_val occur within [l,r]. set l to one beyond last position
            if(last_occ.count(r_val) and l <= last_occ[r_val]){
                sum -= accumulate(nums.begin() + l, nums.begin() + last_occ[r_val] + 1, 0);
                l = last_occ[r_val] + 1;
            }
            last_occ[r_val] = r;
            ans = max(ans, sum);
        }
        return ans;
    }
};

File: array/max_erasure_value.kt
//https://leetcode.com/problems/maximum-erasure-value/
@RequiresApi(Build.VERSION_CODES.N)
fun maximumUniqueSubarray(nums: IntArray): Int {
    var (l, ans, cur_sum) = listOf(0, 0, 0)
    var last_occ = mutableMapOf<Int, Int>()
    for(r in 0 until nums.size){
        cur_sum += nums[r]
        if(last_occ.getOrDefault(nums[r],-1) >= l) {
            var last_val = last_occ.getValue(nums[r])
            while (l <= last_val)
                cur_sum -= nums[l++]
        }
        last_occ[nums[r]] = r
        ans = max(ans, cur_sum)
    }
    return ans
}